### PR TITLE
`ultralytics 8.3.47` fix `softmax` and `Classify` head commonality

### PR DESCRIPTION
**Pull request by Laughing-q**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced YOLO classification support with improved export, processing, and loss handling for better model compatibility and performance. 🚀

### 📊 Key Changes
- Updated version to `8.3.47` for release tracking. ✅
- Added export capability (`export=True`) for the YOLO `Classify` module, enabling export of classification models. 📤
- Improved classification post-processing to handle predictions more efficiently. 📊
- Modified loss function to process tuple-predictions gracefully and calculate classification loss properly. ⚙️
- Adjusted model logic to seamlessly switch between training and inference modes. 📈

### 🎯 Purpose & Impact
- **Enhanced Model Export:** Enables exporting classification models for use in production, improving deployment flexibility. 🌐
- **Improved Usability:** Streamlined classification processing ensures smoother workflows and better usability for developers. 🎉
- **Better Performance:** Refined loss handling and dynamic behavior improve compatibility and accuracy for classification tasks. 📊
- **Future Ready:** Paves the way for improved handling of classification models in YOLO, fostering broader application support. 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Laughing-q, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For your PR:  
- The added `postprocess` method and modifications to the `forward` function look great for improving classification consistency! Please ensure new functionality is thoroughly documented, and any edge cases are covered with appropriate unit tests.  
- If there's a relevant issue in the tracker that this addresses, kindly link it in your PR description. This helps us trace the purpose and impact of the changes effortlessly.  

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). If you have yet to do so, kindly verify that your PR meets the above criteria. An Ultralytics engineer will also assist you soon to review your contribution. 😊  

Thank you for helping to make Ultralytics better! 🚀

**Comment by Laughing-q** [comment]
@glenn-jocher @Y-T-G Hey guys this PR makes our Classify head more consistent with other heads i.e. returning a tuple if `not training` and `not export`, mostly for validation losses, I believe this PR also preserve the validation loss from the scaling issue.
What do you think?
@Y-T-G Can you also test this PR see if everything works? in case I break anything. Thanks!

File: ultralytics/nn/modules/head.py [hunk 1]
```diff
@@ -296,7 +298,10 @@ def forward(self, x):
         if isinstance(x, list):
             x = torch.cat(x, 1)
         x = self.linear(self.drop(self.pool(self.conv(x)).flatten(1)))
-        return x
+        if self.training:
```
Comment (Y-T-G) [inline ultralytics/nn/modules/head.py:301]: @Laughing-q This is how it was before. The problem here is that `self.training` is False during validation, so the validation losses are not correct.

Comment (Y-T-G) [reply ultralytics/nn/modules/head.py:301]: Ah okay, nice.

**Comment by Y-T-G** [comment]
@Laughing-q This works for PyTorch models. The validation losses are correct. And the predictions are also in correct probabilities.
image

```python
In [8]: model = YOLO("yolo11n-cls.pt")

In [9]: results = model.predict(ASSETS / "bus.jpg", imgsz=224)

image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 224x224 minibus 0.57, police_van 0.34, trolleybus 0.04, recreational_vehicle 0.01, streetcar 0.01, 4.7ms
Speed: 8.9ms preprocess, 4.7ms inference, 0.1ms postprocess per image at shape (1, 3, 224, 224)
```

However, the export isn't working as intended. The probabilities are not softmaxed.

```python
In [10]: model = YOLO("yolo11n-cls.onnx")

In [11]: results = model.predict(ASSETS / "bus.jpg", imgsz=224)
Loading yolo11n-cls.onnx for ONNX Runtime inference...
WARNING ⚠️ Failed to start ONNX Runtime session with CUDA. Falling back to CPU...
Preferring ONNX Runtime AzureExecutionProvider

image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 224x224 minibus 14.57, police_van 14.05, trolleybus 11.97, recreational_vehicle 10.88, streetcar 10.00, 24.1ms
Speed: 8.2ms preprocess, 24.1ms inference, 0.2ms postprocess per image at shape (1, 3, 224, 224)
```

**Comment by Y-T-G** [comment]
@Laughing-q My bad. It works. I was using old ONNX export.

```python
In [8]: results = model.predict(ASSETS / "bus.jpg", imgsz=224)
Loading yolo11n-cls.onnx for ONNX Runtime inference...
Preferring ONNX Runtime AzureExecutionProvider

image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 224x224 minibus 0.57, police_van 0.34, trolleybus 0.04, recreational_vehicle 0.01, streetcar 0.01, 15.5ms
Speed: 11.4ms preprocess, 15.5ms inference, 0.2ms postprocess per image at shape (1, 3, 224, 224)
```

**Comment by Laughing-q** [comment]
@Y-T-G ok got it! no worries! I just checked it as well and the softmax layer is in the onnx model.
pic-241207-1846-23

**Comment by glenn-jocher** [comment]
@Laughing-q @Y-T-G this great idea here, we do want to make Classify more consistent and share a more common code base with the other models and dataloaders.

**Review by glenn-jocher — APPROVED** [review]
